src/parser/statement.js

```diff
@@ -861,8 +861,22 @@ pp.parseExport = function (node) {
   return this.finishNode(node, "ExportNamedDeclaration");
 };
 
+pp.shouldParseExportDeclaration = function () {
+  return this.state.type.keyword || this.isContextual("async");
+};
+
+pp.isDeclaration = function (node): boolean {
+  return node.type === "VariableDeclaration"
+    || node.type === "FunctionDeclaration"
+    || node.type === "ClassDeclaration";
+};
+
 pp.parseExportDeclaration = function () {
-  return this.parseStatement(true);
+  const statement = this.parseStatement(true);
+  if (!this.isDeclaration(statement)) {
+    this.unexpected(statement.start);
```

Comment:
After discussing a bit with @hzoo, I changed it to an ahead of time check rather than checking the export statement post-parse. This actually changes the error message in some cases, so wanted to see what do you think. 

If we want a more descriptive error message I could leave the original conditional and do the `shouldParseExportDeclaration` check (maybe now called `isDeclaration`?) inside that block, throwing a more descriptive error if `isDeclaration` returns `false`.

My only concern with the error message is that the intention of the code is. If Babylon is trying to parse `export typeof foo`, can we surmise they intended to export a declaration? Or did you mean a more generic `Not a valid export` type of message? Let me know if I'm missing something!